engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/TimeZoneBranchNode.java

```diff
@@ -0,0 +1,60 @@
+package org.enso.interpreter.node.controlflow.caseexpr;
+
+import com.oracle.truffle.api.RootCallTarget;
+import com.oracle.truffle.api.dsl.Fallback;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.NodeInfo;
+import com.oracle.truffle.api.profiles.ConditionProfile;
+import org.enso.interpreter.runtime.data.EnsoTimeZone;
+import org.enso.interpreter.runtime.data.Type;
+
+/** An implementation of the case expression specialised to working on Time_Zone. */
+@NodeInfo(shortName = "DateMatch")
+public abstract class TimeZoneBranchNode extends BranchNode {
+  private final Type timeZone;
+  private final ConditionProfile profile = ConditionProfile.createCountingProfile();
+
+  TimeZoneBranchNode(Type vector, RootCallTarget branch) {
+    super(branch);
+    this.timeZone = vector;
+  }
+
+  /**
+   * Creates a new node for handling matching on a case expression.
+   *
+   * @param timeZone the expression to use for matching
+   * @param branch the expression to be executed if (@code matcher} matches
+   * @return a node for matching in a case expression
+   */
+  public static TimeZoneBranchNode build(Type timeZone, RootCallTarget branch) {
+    return TimeZoneBranchNodeGen.create(timeZone, branch);
+  }
+
+  @Specialization
+  void doType(VirtualFrame frame, Object state, Type target) {
+    if (profile.profile(timeZone == target)) {
+      accept(frame, state, new Object[0]);
+    }
+  }
+
+  @Specialization
+  void doEnsoZone(VirtualFrame frame, Object state, EnsoTimeZone timeZone) {
+    accept(frame, state, new Object[0]);
+  }
+
+  @Specialization(
+      guards = {"!interop.isDate(zone)", "!interop.isTime(zone)", "interop.isTimeZone(zone)"})
```

Comment:
This is possibly a stupid question, but isn't it better to first have the 'positive' condition and only afterwards the `!` ones?

Seems like in many matches the positive one is more likely to fail first if the thing does not match, so the short circuit will rule out that match faster. Of course it may be a premature micro optimization, so I'm not sure if I should even be mentioning it. But intuition suggests it may be better for almost no cost - and also it seems a bit more readable to me - I first check _what I want this to be_ and only then make sure it _is not also something else_.

Reply:
My preference is the opposite but I also want to have that PR merged so I will make the change :)